Code/Engine/Foundation/Platform/Win/DirectoryWatcher_Win.cpp

```diff
@@ -36,13 +39,246 @@ namespace
+            }
+            else
+            {
+              lastChangeAtPath[currentChange.eventFilePath] = i;
+            }
+            lastMoveFrom = i;
+            break;
+          case FILE_ACTION_RENAMED_NEW_NAME:
+            EZ_ASSERT_DEBUG(lastMoveFrom != -1, "last move from should be present when encountering FILE_ACTION_RENAMED_NEW_NAME");
+            nextOp[lastMoveFrom] = i;
+            lastChangeAtPath.Remove(ref_changes[lastMoveFrom].eventFilePath);
+            lastChangeAtPath.Insert(currentChange.eventFilePath, i);
+            break;
+        }
+      }
+    }
+
+    // Anything that is chained via the nextOp linked list must be given the same type.
+    // Instead of building arrays of arrays, we create a bit field of all changes and then flatten the linked list at the first set bit. While iterating we remove everything we reached via the linked list so on the next call to get the first bit we will find another object that needs processing. As the operations are ordered, the first bit will always point to the very first operation of an object (nextOp can never point to a previous element).
+    ezBitfield<ezHybridArray<ezUInt32, 4>> pendingChanges;
+    pendingChanges.SetCount(ref_changes.GetCount(), true);
+
+    // Get start of first object.
+    ezHybridArray<Change*, 4> objectChanges;
+    auto it = pendingChanges.GetIterator();
+    while (it.IsValid())
+    {
+      // Flatten the changes for one object into a list for easier processing.
+      {
+        objectChanges.Clear();
+        ezUInt32 currentIndex = it.Value();
+        objectChanges.PushBack(&ref_changes[currentIndex]);
+        pendingChanges.ClearBit(currentIndex);
+        while (nextOp[currentIndex] != -1)
+        {
+          currentIndex = nextOp[currentIndex];
+          pendingChanges.ClearBit(currentIndex);
+          objectChanges.PushBack(&ref_changes[currentIndex]);
+        }
+      }
+
+      // Figure out what type the object is. There is no correct way of doing this, which is the reason why ReadDirectoryChangesExW exists. There are however some good heuristics we can use:
+      // 1. If the change is on an existing object, we should know it's type from the mirror. This is 100% correct.
+      // 2. If object still exists, we can query its stats on disk to determine its type. In rare cases, this can be wrong but you would need to create a case where a file is replaced with a folder of the same name or something.
+      // 3. If the object was created and deleted in the same enumeration, we cannot know what type it was, so we take a guess.
+      {
+        bool isFile = true;
+        bool typeFound = false;
+        for (Change* currentChange : objectChanges)
+        {
+          ezFileSystemMirrorType::Type type;
+          if (pMirror->GetType(currentChange->eventFilePath, type).Succeeded())
+          {
+            isFile = type == ezFileSystemMirrorType::Type::File;
+            typeFound = true;
+            break;
```

Comment:
goto typeFound; 😄

Reply:
That would work as well, but I'm old fashioned and have a goto phobia (or rather I forgot how it works).